File: Java/src/main/java/drawing/shapes/Shape.java
package drawing.shapes;

import drawing.formatting.Formatter;
import drawing.writing.JPEGWriter;
import drawing.writing.PNGWriter;

import java.io.IOException;
import java.io.Writer;

public interface Shape {

    Line[] toLines();

    default void draw(Writer writer, Formatter formatter) throws IOException {
        // Moved 'toLines' call here (out of Drawing).
        Line[] lines = this.toLines();
        for (Line line : lines) {
            writer.write(formatter.format(line));
        }
    }
}

File: TypeScript/src/drawing/shapes/shape.ts
import { Writer } from "../writing/writer";
import { PNGWriter } from "../writing/pngwriter";
import { JPEGWriter } from "../writing/jpegwriter";
import { Line } from "./line";
import { Formatter } from "../formatting/formatter";

interface Shape {
	
    toLines(): Line[];

    draw(writer: Writer, formatter: Formatter): void;
}

abstract class AbstractShape implements Shape {

	abstract toLines(): Line[];

	public draw(writer: Writer, formatter: Formatter): void {
		let lines: Line[] = this.toLines();
        for (let line of lines) {
            writer.write(formatter.format(line));
        }
	}
}

export { Shape, AbstractShape }